File: app/lib/google-maps.ts
import type { VenueSearchParams, GoogleMapsGeocodingResult } from '@/types/venue';

/**
 * Geocode a venue using Google Maps Geocoding API
 * Handles misspellings and partial matches
 */
export async function geocodeVenue(
  params: VenueSearchParams
): Promise<GoogleMapsGeocodingResult[]> {
  // Build address string from venue params
  const addressParts = [
    params.name,
    params.city,
    params.state,
    params.country
  ].filter(Boolean);
  
  const address = addressParts.join(', ');
  
  const apiKey = process.env.GOOGLE_MAPS_API_KEY;
  if (!apiKey) {
    throw new Error('GOOGLE_MAPS_API_KEY not configured');
  }
  
  const url = new URL('https://maps.googleapis.com/maps/api/geocode/json');
  url.searchParams.set('address', address);
  url.searchParams.set('key', apiKey);
  
  try {
    const response = await fetch(url.toString(), {
      signal: AbortSignal.timeout(10000), // 10s timeout
    });
    
    if (!response.ok) {
      console.error(`Google Geocoding API error: ${response.status}`);
      return [];
    }
    
    const data = await response.json();
    
    if (data.status !== 'OK' && data.status !== 'ZERO_RESULTS') {
      console.error(`Geocoding API status: ${data.status}`, data.error_message);
      return [];
    }
    
    return data.results || [];
  } catch (error) {
    console.error('Error geocoding venue:', error);
    return [];
  }
}

/**
 * Extract city from address components
 */
export function extractCity(addressComponents: GoogleMapsGeocodingResult['address_components']): string | null {
  const city = addressComponents.find(c => 
    c.types.includes('locality') || 
    c.types.includes('sublocality')
  );
  return city?.long_name || null;
}

/**
 * Extract state from address components
 */
export function extractState(addressComponents: GoogleMapsGeocodingResult['address_components']): string | null {
  const state = addressComponents.find(c => 
    c.types.includes('administrative_area_level_1')
  );
  return state?.short_name || state?.long_name || null;
}

/**
 * Extract country from address components
 */
export function extractCountry(addressComponents: GoogleMapsGeocodingResult['address_components']): string {
  const country = addressComponents.find(c => c.types.includes('country'));
  return country?.long_name || 'Unknown';
}

File: services/agent/app/clients/google_maps.py
from typing import Any

import httpx

from app.config import settings


class GeocodingResult:
    """Typed wrapper around Google Maps Geocoding API result."""

    def __init__(self, data: dict[str, Any]) -> None:
        self._data = data

    @property
    def place_id(self) -> str:
        return str(self._data.get("place_id", ""))

    @property
    def formatted_address(self) -> str:
        return str(self._data.get("formatted_address", ""))

    @property
    def latitude(self) -> float:
        return float(self._data["geometry"]["location"]["lat"])

    @property
    def longitude(self) -> float:
        return float(self._data["geometry"]["location"]["lng"])

    @property
    def address_components(self) -> list[dict[str, Any]]:
        return list(self._data.get("address_components", []))

    @property
    def partial_match(self) -> bool:
        return bool(self._data.get("partial_match", False))


async def geocode_venue(
    name: str,
    city: str | None = None,
    state: str | None = None,
    country: str | None = None,
) -> list[GeocodingResult]:
    """Geocode a venue using Google Maps Geocoding API."""
    address_parts = [p for p in [name, city, state, country] if p]
    address = ", ".join(address_parts)

    api_key = settings.google_maps_api_key
    if not api_key:
        return []

    async with httpx.AsyncClient(timeout=10.0) as client:
        response = await client.get(
            "https://maps.googleapis.com/maps/api/geocode/json",
            params={"address": address, "key": api_key},
        )

    if response.status_code != 200:
        return []

    data = response.json()
    if data.get("status") not in ("OK", "ZERO_RESULTS"):
        return []

    return [GeocodingResult(r) for r in data.get("results", [])]


def extract_city(address_components: list[dict[str, Any]]) -> str | None:
    for component in address_components:
        types = component.get("types", [])
        if "locality" in types or "sublocality" in types:
            return str(component.get("long_name"))
    return None


def extract_state(address_components: list[dict[str, Any]]) -> str | None:
    for component in address_components:
        types = component.get("types", [])
        if "administrative_area_level_1" in types:
            return str(component.get("short_name") or component.get("long_name"))
    return None


def extract_country(address_components: list[dict[str, Any]]) -> str:
    for component in address_components:
        types = component.get("types", [])
        if "country" in types:
            return str(component.get("long_name", "Unknown"))
    return "Unknown"
